[Bug] worker-loader cannot work with umi@3.4.0

## What happens?

web worker cannot build successfully with umi@3.4.0.

worker-loader cannot find dependent module

Error: Cannot find module 'webpack/lib/webworker/WebWorkerTemplatePlugin'

## Mini Showcase Repository(REQUIRED)

https://github.com/Leemin0311/umi-worder-loader-bug

## How To Reproduce

**Steps to reproduce the behavior:** 
1. yarn
2. yarn start

**Expected behavior** 
1. build successfully

## Context

- **Umi Version**: 3.4.0
- **Node Version**: 14.7.0
- **Platform**: MacOS 11.2.2

--- a/docs/config/README.md
+++ b/docs/config/README.md
@@ -1249,3 +1249,10 @@ export default {
 包含以下子配置项：
 
 - lazyCompilation，是否启用基于路由的按需编译
+
+## workerLoader <Badge>3.4.1+</Badge>
+
+- Type: `object`
+- Default: `false`
+
+开启 worker-loader 功能。

--- a/docs/config/README.zh-CN.md
+++ b/docs/config/README.zh-CN.md
@@ -1248,3 +1248,10 @@ export default {
 包含以下子配置项：
 
 - lazyCompilation，是否启用基于路由的按需编译
+
+## workerLoader <Badge>3.4.1+</Badge>
+
+- Type: `object`
+- Default: `false`
+
+开启 worker-loader 功能。

--- a/packages/bundler-webpack/package.json
+++ b/packages/bundler-webpack/package.json
@@ -34,7 +34,7 @@
   },
   "dependencies": {
     "@umijs/bundler-utils": "3.4.0",
-    "@umijs/deps": "0.2.34",
+    "@umijs/deps": "0.3.1",
     "@umijs/types": "3.4.0",
     "@umijs/utils": "3.4.0",
     "jest-worker": "26.6.2",

--- a/packages/bundler-webpack/package.json
+++ b/packages/bundler-webpack/package.json
@@ -34,7 +34,7 @@
   },
   "dependencies": {
     "@umijs/bundler-utils": "3.4.0",
-    "@umijs/deps": "0.3.1",
+    "@umijs/deps": "0.3.3",
     "@umijs/types": "3.4.0",
     "@umijs/utils": "3.4.0",
     "jest-worker": "26.6.2",

--- a/packages/bundler-webpack/src/fixtures/worker-loader/config.ts
+++ b/packages/bundler-webpack/src/fixtures/worker-loader/config.ts
@@ -1,0 +1,3 @@
+export default {
+  workerLoader: {},
+}

--- a/packages/bundler-webpack/src/fixtures/worker-loader/expect.ts
+++ b/packages/bundler-webpack/src/fixtures/worker-loader/expect.ts
@@ -1,0 +1,7 @@
+import { IExpectOpts } from '../types';
+
+export default ({ indexJS, files }: IExpectOpts) => {
+  expect(files).toContain('index.js');
+  expect(files).toContain('worker.worker.js');
+  expect(indexJS).toContain(`return new Worker(__webpack_require__.p + "worker.worker.js");`);
+}

--- a/packages/bundler-webpack/src/fixtures/worker-loader/index.ts
+++ b/packages/bundler-webpack/src/fixtures/worker-loader/index.ts
@@ -1,0 +1,4 @@
+import Worker from './worker';
+
+const worker = new Worker();
+worker.postMessage('aaaa');

--- a/packages/bundler-webpack/src/fixtures/worker-loader/worker.js
+++ b/packages/bundler-webpack/src/fixtures/worker-loader/worker.js
@@ -1,0 +1,4 @@
+
+self.addEventListener('message', (e) => {
+  console.log('message received');
+});

--- a/packages/bundler-webpack/src/getConfig/getConfig.ts
+++ b/packages/bundler-webpack/src/getConfig/getConfig.ts
@@ -352,6 +352,16 @@ export default async function getConfig(
     .use('raw-loader')
       .loader(require.resolve('@umijs/deps/compiled/raw-loader'));
 
+  if (config.workerLoader) {
+    // prettier-ignore
+    webpackConfig.module
+      .rule('worker')
+      .test(/.*worker.(ts|js)/)
+      .use('worker-loader')
+        .loader(require.resolve('@umijs/deps/compiled/worker-loader'))
+        .options(config.workerLoader);
+  }
+
   // css
   css({
     type,

--- a/packages/core/package.json
+++ b/packages/core/package.json
@@ -26,7 +26,7 @@
   "dependencies": {
     "@umijs/ast": "3.4.0",
     "@umijs/babel-preset-umi": "3.4.0",
-    "@umijs/deps": "0.2.34",
+    "@umijs/deps": "0.3.1",
     "@umijs/utils": "3.4.0"
   }
 }

--- a/packages/core/package.json
+++ b/packages/core/package.json
@@ -26,7 +26,7 @@
   "dependencies": {
     "@umijs/ast": "3.4.0",
     "@umijs/babel-preset-umi": "3.4.0",
-    "@umijs/deps": "0.3.1",
+    "@umijs/deps": "0.3.3",
     "@umijs/utils": "3.4.0"
   }
 }

--- a/packages/preset-built-in/package.json
+++ b/packages/preset-built-in/package.json
@@ -28,7 +28,7 @@
     "@types/react-router-config": "5.0.1",
     "@umijs/babel-preset-umi": "3.4.0",
     "@umijs/bundler-webpack": "3.4.0",
-    "@umijs/deps": "0.3.1",
+    "@umijs/deps": "0.3.3",
     "@umijs/renderer-mpa": "3.4.0",
     "@umijs/renderer-react": "3.4.0",
     "@umijs/runtime": "3.4.0",

--- a/packages/preset-built-in/package.json
+++ b/packages/preset-built-in/package.json
@@ -28,7 +28,7 @@
     "@types/react-router-config": "5.0.1",
     "@umijs/babel-preset-umi": "3.4.0",
     "@umijs/bundler-webpack": "3.4.0",
-    "@umijs/deps": "0.2.34",
+    "@umijs/deps": "0.3.1",
     "@umijs/renderer-mpa": "3.4.0",
     "@umijs/renderer-react": "3.4.0",
     "@umijs/runtime": "3.4.0",

--- a/packages/preset-built-in/src/index.ts
+++ b/packages/preset-built-in/src/index.ts
@@ -67,6 +67,7 @@ export default function () {
       require.resolve('./plugins/features/umiInfo'),
       require.resolve('./plugins/features/runtimeHistory'),
       require.resolve('./plugins/features/webpack5'),
+      require.resolve('./plugins/features/workerLoader'),
 
       // html
       require.resolve('./plugins/features/html/favicon'),

--- a/packages/preset-built-in/src/plugins/features/workerLoader.ts
+++ b/packages/preset-built-in/src/plugins/features/workerLoader.ts
@@ -1,0 +1,12 @@
+import { IApi } from '@umijs/types';
+
+export default (api: IApi) => {
+  api.describe({
+    key: 'workerLoader',
+    config: {
+      schema(joi) {
+        return joi.object();
+      },
+    },
+  });
+};

--- a/packages/server/package.json
+++ b/packages/server/package.json
@@ -24,7 +24,7 @@
     "access": "public"
   },
   "dependencies": {
-    "@umijs/deps": "0.2.34",
+    "@umijs/deps": "0.3.1",
     "@umijs/utils": "3.4.0"
   }
 }

--- a/packages/server/package.json
+++ b/packages/server/package.json
@@ -24,7 +24,7 @@
     "access": "public"
   },
   "dependencies": {
-    "@umijs/deps": "0.3.1",
+    "@umijs/deps": "0.3.3",
     "@umijs/utils": "3.4.0"
   }
 }

--- a/packages/types/package.json
+++ b/packages/types/package.json
@@ -27,7 +27,7 @@
   "dependencies": {
     "@umijs/babel-preset-umi": "3.4.0",
     "@umijs/core": "3.4.0",
-    "@umijs/deps": "0.2.34",
+    "@umijs/deps": "0.3.1",
     "@umijs/server": "3.4.0",
     "@umijs/utils": "3.4.0",
     "webpack-chain": "6.5.1"

--- a/packages/types/package.json
+++ b/packages/types/package.json
@@ -27,7 +27,7 @@
   "dependencies": {
     "@umijs/babel-preset-umi": "3.4.0",
     "@umijs/core": "3.4.0",
-    "@umijs/deps": "0.3.1",
+    "@umijs/deps": "0.3.3",
     "@umijs/server": "3.4.0",
     "@umijs/utils": "3.4.0",
     "webpack-chain": "6.5.1"

--- a/packages/umi/package.json
+++ b/packages/umi/package.json
@@ -30,7 +30,7 @@
   "dependencies": {
     "@umijs/bundler-webpack": "3.4.0",
     "@umijs/core": "3.4.0",
-    "@umijs/deps": "0.2.34",
+    "@umijs/deps": "0.3.1",
     "@umijs/preset-built-in": "3.4.0",
     "@umijs/runtime": "3.4.0",
     "@umijs/types": "3.4.0",

--- a/packages/umi/package.json
+++ b/packages/umi/package.json
@@ -30,7 +30,7 @@
   "dependencies": {
     "@umijs/bundler-webpack": "3.4.0",
     "@umijs/core": "3.4.0",
-    "@umijs/deps": "0.3.1",
+    "@umijs/deps": "0.3.3",
     "@umijs/preset-built-in": "3.4.0",
     "@umijs/runtime": "3.4.0",
     "@umijs/types": "3.4.0",

--- a/packages/utils/package.json
+++ b/packages/utils/package.json
@@ -24,6 +24,6 @@
     "access": "public"
   },
   "dependencies": {
-    "@umijs/deps": "0.2.34"
+    "@umijs/deps": "0.3.1"
   }
 }

--- a/packages/utils/package.json
+++ b/packages/utils/package.json
@@ -24,6 +24,6 @@
     "access": "public"
   },
   "dependencies": {
-    "@umijs/deps": "0.3.1"
+    "@umijs/deps": "0.3.3"
   }
 }

